integr8ly/integreatly-operator
# [INTLY-2813] Clean up oauth clients on cr deletion

### Description
Cleans up oauth clients and other cluster level objects when an `Installation` custom resource is deleted. This is done with the use of a new finalizer reconciler which accepts an anonymous function that handles the execution of the finalization logic. 

Each product implements its own finalizer and handles its own finalization logic and the deletion of its finalizer from the CR.

### Verification steps

- Create the operator source which points to my app registry:
```yml
apiVersion: operators.coreos.com/v1
kind: OperatorSource
metadata:
  finalizers:
    - finalizer.operatorsources.operators.coreos.com
  name: integreatly-operators
  namespace: openshift-marketplace
spec:
  authorizationToken: {}
  displayName: Integreatly Operators
  endpoint: 'https://quay.io/cnr'
  publisher: Integreatly Publisher
  registryNamespace: dimitra
  type: appregistry
```
- Create a new namespace to test in, and create the secrets required for 3scale and launcher
- Create a subscription to the Integreatly operator from the Operator Hub in the namespace just created, then create an installation custom resource
```yml
apiVersion: integreatly.org/v1alpha1
kind: Installation
metadata:
  name: example-installation
spec:
  type: workshop
  namespacePrefix: integreatly-
  routingSubdomain: apps.cluster-pbrookes-1b0d.pbrookes-1b0d.openshiftworkshop.com
  masterUrl: https://console-openshift-console.apps.cluster-pbrookes-1b0d.pbrookes-1b0d.openshiftworkshop.com
  selfSignedCerts: true
```
- Once all the products have completed, verify that 3 new finalizers have been added to the custom resource yaml 
- Delete the installation. Verify that each product finalizer (`finalizer.rhsso.integreatly.org`, `finalizer.3scale.integreatly.org`, `finalizer.webapp.integreatly.org`) is removed from the cr once its corresponding oauth client has been deleted (verify that the following oauth clients are successfully cleaned up: `rhsso`, `3scale`, `integreatly-solution-explorer`)
- Verify that once these finalizers have been removed from the installation cr, the installation itself is also deleted and the product namespaces are cleaned up
